datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/micrometer/DataSourceMetrics.java

```diff
@@ -0,0 +1,106 @@
+package com.github.gavlyukovskiy.micrometer;
+
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.stats.hist.Histogram;
+import io.micrometer.core.instrument.stats.quantile.WindowSketchQuantiles;
+import org.springframework.boot.autoconfigure.jdbc.metadata.DataSourcePoolMetadata;
+
+import javax.sql.DataSource;
+
+import java.sql.Connection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * {@link DataSource} metrics linked to micrometer's {@link MeterRegistry}.
+ *
+ * @author Arthur Gavlyukovskiy
+ * @since 1.2.2
+ */
+public class DataSourceMetrics {
```

Comment:
Recommend to implement `MeterBinder`, as then if you wire this class as a `@Bean` in Spring, all its metrics are automatically bound to the registry defined to the Spring context.

Reply:
The problem is that I may have multiple `DataSource` beans in the context, this class is responsive for the specific `DataSource` bean and managed by one of proxy provider specific listeners. Even if I make it a bean I will need a metric holder for concrete `DataSource`. But it's a better approach to register metrics for all data sources in the context rather than create one once fist connection is retrieved.

Reply:
In spring you could just wire each MeterBinder with a different qualifer

Reply:
Or just accept a list of data source and handle them all here?

Reply:
Yes, this is what I've done at the end. Thanks.